src/libs/print_settings.c

```diff
@@ -0,0 +1,1398 @@
+    for(int y = 0; y < d->height; y++)
+    {
+      for(int x = 0; x < d->width; x++, in_ptr += 4, out_ptr += 3)
+        memcpy(out_ptr, in_ptr, 3);
+    }
+  }
+  else
+  {
+    const uint16_t *in_ptr = (const uint16_t *)in;
+    uint16_t *out_ptr = (uint16_t *)d->ps->buf;
+    for(int y = 0; y < d->height; y++)
+    {
+      for(int x = 0; x < d->width; x++, in_ptr += 4, out_ptr += 3)
+        memcpy(out_ptr, in_ptr, 6);
+    }
+  }
+
+  return 0;
+}
+
+static void
+_print_button_clicked (GtkWidget *widget, gpointer user_data)
+{
+  dt_lib_module_t *self = (dt_lib_module_t *)user_data;
+  dt_lib_print_settings_t *ps = (dt_lib_print_settings_t *)self->data;
+
+  const int imgid = dt_view_filmstrip_get_activated_imgid(darktable.view_manager);
+
+  if (imgid == -1)
+  {
+    dt_control_log(_("cannot print until a picture is selected"));
+    dt_control_queue_redraw();
+    return;
+  }
+  if (strlen(ps->prt.printer.name) == 0 || ps->prt.printer.resolution == 0)
+  {
+    dt_control_log(_("cannot print until a printer is selected"));
+    dt_control_queue_redraw();
+    return;
+  }
+  if (ps->prt.paper.width == 0 || ps->prt.paper.height == 0)
+  {
+    dt_control_log(_("cannot print until a paper is selected"));
+    dt_control_queue_redraw();
+    return;
+  }
+
+  dt_control_log(_("prepare printing image %d on `%s'"), imgid, ps->prt.printer.name);
+  dt_control_queue_redraw();
+
+  // user margin are already in the proper orientation landscape/portrait
+  double width, height;
+  double margin_w = ps->prt.page.margin_left + ps->prt.page.margin_right;
+  double margin_h = ps->prt.page.margin_top + ps->prt.page.margin_bottom;
+
+  if (ps->prt.page.landscape)
+  {
+    width = ps->prt.paper.height;
+    height = ps->prt.paper.width;
+    margin_w += ps->prt.printer.hw_margin_top + ps->prt.printer.hw_margin_bottom;
+    margin_h += ps->prt.printer.hw_margin_left + ps->prt.printer.hw_margin_right;
+  }
+  else
+  {
+    width = ps->prt.paper.width;
+    height = ps->prt.paper.height;
+    margin_w += ps->prt.printer.hw_margin_left + ps->prt.printer.hw_margin_right;
+    margin_h += ps->prt.printer.hw_margin_top + ps->prt.printer.hw_margin_bottom;
+  }
+
+  const int32_t width_pix = (width * ps->prt.printer.resolution) / 25.4;
+  const int32_t height_pix = (height * ps->prt.printer.resolution) / 25.4;
+
+  const double pa_width  = (width  - margin_w) / 25.4;
+  const double pa_height = (height - margin_h) / 25.4;
+
+  if (darktable.unmuted & DT_DEBUG_PRINT)
+    printf("[print] area for image %u : %3.2fin x %3.2fin\n", imgid, pa_width, pa_height);
+
+  // compute the needed size for picture for the given printer resolution
+
+  int max_width  = (pa_width  * ps->prt.printer.resolution);
+  int max_height = (pa_height * ps->prt.printer.resolution);
+
+  // make sure we are a multiple of four
+
+  max_width -= max_width % 4;
+  max_height -= max_height % 4;
+
+  const gchar *printer_profile = dt_conf_get_string("plugins/print/printer/iccprofile");
```

Comment:
I would prefer if those would be handed over in self. I know that we go throught he conf in many places, but that is not really a good design.